docs/headers/fmi3LsBusCan.h

```diff
@@ -0,0 +1,375 @@
+*/
+
+#include "fmi3lsBus.h"
+#include <assert.h>
+
+/***************************************************
+  CAN bus specific OP code 
+****************************************************/
+
+/**
+ * Initiates the transmission of CAN frames.
+ */
+#define FMI3_LS_BUS_CAN_OP_CAN_TRANSMIT ((fmi3LsBusOperationType)0x0010)
+
+/**
+ * Represents an operation for the transmission of a CAN FD frame.
+ */
+#define FMI3_LS_BUS_CAN_OP_CANFD_TRANSMIT ((fmi3LsBusOperationType)0x0011)
+
+/**
+ * Represents an operation for the transmission of a CAN XL frame.
+ */
+#define FMI3_LS_BUS_CAN_OP_CANXL_TRANSMIT ((fmi3LsBusOperationType)0x0012)
+
+/**
+ * Signals successful receipt of a transmitted CAN, CAN FD and CAN XL frame to simulate a CAN acknowledgment behavior.
+ */
+#define FMI3_LS_BUS_CAN_OP_CONFIRM ((fmi3LsBusOperationType)0x0020)
+
+/**
+ * The Arbitration Lost operation indicates that a CAN frame could not be sent immediately and was therefore
+ * discarded by the Bus Simulation.
+ */
+#define FMI3_LS_BUS_CAN_OP_ARBITRATION_LOST ((fmi3LsBusOperationType)0x0030)
+
+/**
+ * Represents an operation for simulated bus errors.
+ */
+#define FMI3_LS_BUS_CAN_OP_CAN_BUS_ERROR ((fmi3LsBusOperationType)0x0031)
+
+/**
+ * Represents an operation for the configuration of a Bus Simulation. In detail, the configuration of a CAN, CAN FD
+ * and CAN XL baud rate is possible. Also the configuration of further options, like buffer handling, is supported
+ * by this operation.
+ */
+#define FMI3_LS_BUS_CAN_OP_CONFIGURATION ((fmi3LsBusOperationType)0x0040)
+
+/**
+ * FMI virtual bus operation which indicates a status operation.
+ */
+#define FMI3_LS_BUS_CAN_OP_STATUS ((fmi3LsBusOperationType)0x0041)
+
+/**
+ * Represents an operation for triggering a bus-specific wake up.
+ */
+#define FMI3_LS_BUS_CAN_OP_WAKEUP ((fmi3LsBusOperationType)0x0042)
+
+
+
+/***************************************************
+Types for CAN specific bus operations
+****************************************************/
+
+#pragma pack(1)
+
+/**
+ * Data type representing the CAN identifer.
+ */
+typedef fmi3UInt32 fmi3LsBusCanId;
+
+/**
+ * Data type for defining a standard or extended messages.
+ */
+typedef fmi3Boolean fmi3LsBusCanIde;
+
+/**
+ * Data type for defining a Remote Transmission Request frame.
+ */
+typedef fmi3Boolean fmi3LsBusCanRtr;
+
+/**
+ * Data type for data length.
+ */
+typedef fmi3UInt16 fmi3LsBusCanDataLength;
+
+/**
+ * Data type for data.
+ */
+typedef fmi3UInt8 fmi3LsBusCanData;
+
+/**
+ * Can transmit operation.
+ */
+typedef struct
+{
+    fmi3LsBusOperationHeader header; /**< Message header */
+    fmi3LsBusCanId  id; /**<  CAN message ID. */
+    fmi3LsBusCanIde ide; /**< Standard (11-bit) or Extended (29-bit) message identifier. */
+    fmi3LsBusCanRtr rtr; /**< Remote Transmission Request frame. */
+    fmi3LsBusCanDataLength dataLength; /**< Data length. */
+    fmi3LsBusCanData data[]; /**< Data. */
```

Comment:
Visual Studio blames about Null-Array (C4200). Should it be fmi2LsBusCanData* ?

Reply:
I agree with you completely here @PenndorfT. I'll see how I can solve the problem and get back to you.

Reply:
@PenndorfT I fixed it. I cannot use a pointer in that case because doing this will cause that the struct cannot be easily transformed to an operation using memcopy(...). Is is OK for you? 

PS: I will close the other two same comments and we can discuss the solution within this comment.

Reply:
Hi @PenndorfT, @bmenne-dspace,

I would like to reopen the discussion about the Visual Studio Null-Array (C4200) warning.

The intention of the type-definition like fmi3LsBusCanOperationCanTransmit is to represent the received data in such a way, that it can be used to cast the received binary data into the specific operation. In this sense, the type must represent the operation exactly as define in the specification. The “data” element of fmi3LsBusCanOperationCanTransmit is therefore define as a "flexible array member” because its actual length is only known at runtime.
Flexible array members (https://en.wikipedia.org/wiki/Flexible_array_member) were introduced with C99, but until now they are not officially supported by C++. Nevertheless, the most important C++ compilers (clang, gcc, msvc) supporting “Flexible array members”, because there is no good alternative. 

If we define the data element to be a point for example (fmi3LsBusCanData* data;), then it would not represent the operation as defined in the specification. Simple casting of the binary data received by the RX_Data variable would therefore not possible any longer. Keep in mind that the "data" element is an embedded array not a pointer.  

Defining an “Array-Size” for example (fmi3LsBusCanData data[MAX_DATA_LENGTH];) would be a workaround, but in this case, the size should be the maximum length, not the minimum length of 1 as it is defined currently. When defining the minimum length, then a compiler might submits array bounding check warnings. In either way, a sizeof() would not deliver a useful value. 

By suggestion therefore would be to keep using flexible array member for the data elements and add some comments in the specification about the header files and their C conformance:

Proposal:
Equal to the FMI API, the FMI-LS-BUS header files are written in C, not C++. Additionally, the C compiler shall support the C11 standard since static assertions have been used to ensure correct struct packings. If the header will be compiled with a C++ compiler, then the compiler shall support flexible array struct members as defined by the C99 standard. Most popular C++ compiler (clang, gcc, msvc) supporting flexible array members, but it could be necessary to use additional compiler options. 

For example: When using the msvc compiler the option /Za to “Disable Language extensions” must not be used. Additionally the warning C4200 “nonstandard extension used: zero-sized array in struct/union” can be suppressed by using the compiler option /wd4200.

Reply:
Thanks, I fixed it.